File: Calendar.java
package DevProject;
import java.time.LocalDate;
import java.time.YearMonth;
import java.time.DayOfWeek;
import java.util.ArrayList;
import java.util.List;

public class Calendar {
	// Creates a new list of activities and stores them in an array list
	private List<Activity> activities;

	public Calendar() {
		activities = new ArrayList<>();

	}

	public static void main(String args[]) {
		Calendar calendar = new Calendar();

		// Testing to simulate user input
		calendar.createActivity("Create Login Page", 10, 5, 2024, 10, 15, 2024);
		calendar.createActivity("Test", 11, 11, 2024, 11, 11, 2024);

		calendar.displayCalendar(2024); // Displays the entire year of 2024
		calendar.displayCalendar(2024, 10); // Displays only October of 2024


	}


	public void createActivity(String name, int startMonth, int startDay, int startYear, int endMonth, int endDay, int endYear) {
		LocalDate startDate = LocalDate.of(startYear, startMonth, startDay);
		LocalDate deadLine = LocalDate.of(endYear, endMonth, endDay);
		Activity activity = new Activity(name, startDate, deadLine);
		activities.add(activity);

	}

	private void displayCalendar(int year, int month) {
		printMonth(year, month);
		printActivitiesInMonth(year, month);

	}

	private void displayCalendar(int year) {
		for (int month = 1; month <= 12; month++) {
			displayCalendar(year, month);

		}
	}

	public static void printMonth(int year, int month) {
		// Gets the number of days in the current month based on the current date
		YearMonth yearMonth = YearMonth.of(year, month);
		int numberOfDaysInMonth = yearMonth.lengthOfMonth();

		// Testing to see if the current number of days in the month is correct
		// System.out.print("The number of days in the current month are: " + numberOfDaysInMonth);

		// Gets the day that the month starts on
		LocalDate firstDayOfMonth = yearMonth.atDay(1);
		DayOfWeek startDay = firstDayOfMonth.getDayOfWeek();

		// Testing to see if the day that the month starts on the correct day
		// System.out.print("The current month starts on: " + startDay);

		// Header for the current month
		System.out.println("\n " + yearMonth.getMonth() + " " + year);
		System.out.println("Su Mo Tu We Th Fr Sa");

		// Determines how much space is given at the start of the month for the first day
		int dayOfWeekSpace = startDay.getValue();
		int startingDay = (dayOfWeekSpace % 7);
		// Prints out the initial spaces at the start of the month
		for (int i = 0; i < startingDay; i++)
			System.out.print("   ");

		// Prints out the days(numbers) of the month
		for (int i = 1; i <= numberOfDaysInMonth; i++) {
			System.out.printf("%2d ", i);
			if ((i + startingDay) % 7 == 0)
				System.out.println();

		}
		System.out.println();

	}

	private void printActivitiesInMonth(int year, int month) {
		List<Activity> thisMonthsActivities = getActivitiesInMonth(year, month);
		System.out.println("\nTasks and Projects for " + YearMonth.of(year, month).getMonth() + ":");
		for (Activity activity : thisMonthsActivities) {
			System.out.println(" - " + activity.getName() + 
					" | Starting Date: " + (activity.getActivityStart() != null ? activity.getActivityStart() : "N/A") + 
					" | Deadline: " + (activity.getDeadline() != null ? activity.getDeadline() : "N/A"));

		}

	}


	public List<Activity> getActivitiesInMonth(int year, int month) {
		List<Activity> thisMonthsActivities = new ArrayList<>();
		for (Activity activity : activities) {
			LocalDate startDate = activity.getActivityStart();
			LocalDate deadLine = activity.getDeadline();
			if ((startDate != null && startDate.getYear() == year && startDate.getMonthValue() == month) ||
					(deadLine != null && deadLine.getYear() == year && deadLine.getMonthValue() == month)) {
				thisMonthsActivities.add(activity);

			}
		}
		return thisMonthsActivities;

	}

}

File: Calendar.js
class Calendar {
    constructor() {
      // Stores activities for the calendar
      this.activities = [];
    }
  
    // Creates an activity with a name and start/end dates
    createActivity(name, startMonth, startDay, startYear, endMonth, endDay, endYear) {
      const startDate = new Date(startYear, startMonth - 1, startDay); // Months are 0-indexed in JavaScript Date
      const endDate = new Date(endYear, endMonth - 1, endDay);
      const activity = new Activity(name, startDate, endDate);
      this.activities.push(activity);
    }
  
    // Display the calendar for a specific year and month
    displayCalendar(year, month) {
      this.printMonth(year, month);
      this.printActivitiesInMonth(year, month);
    }
  
    // Display the entire year
    displayCalendarYear(year) {
      for (let month = 1; month <= 12; month++) {
        this.displayCalendar(year, month);
      }
    }
  
    // Print the month in the calendar
    printMonth(year, month) {
      const date = new Date(year, month - 1, 1); // Start of the month
      const numberOfDaysInMonth = new Date(year, month, 0).getDate(); // Number of days in the month
      const startDay = date.getDay(); // The first day of the week (0-Sunday, 1-Monday, etc.)
  
      console.log(`\n ${date.toLocaleString('default', { month: 'long' })} ${year}`);
      console.log('Su Mo Tu We Th Fr Sa');
  
      // Add leading spaces for the first week
      let space = '   '.repeat(startDay);
      process.stdout.write(space);
  
      // Print the days of the month
      for (let day = 1; day <= numberOfDaysInMonth; day++) {
        process.stdout.write(`${String(day).padStart(2, ' ')} `);
        if ((day + startDay) % 7 === 0) {
          console.log(); // New line after each week
        }
      }
      console.log(); // Ensure the final line ends properly
    }
  
    // Print activities for the specified month
    printActivitiesInMonth(year, month) {
      const activitiesInMonth = this.getActivitiesInMonth(year, month);
      console.log(`\nTasks and Projects for ${new Date(year, month - 1).toLocaleString('default', { month: 'long' })}:`);
      activitiesInMonth.forEach(activity => {
        console.log(` - ${activity.name} | Starting Date: ${activity.activityStart.toDateString()} | Deadline: ${activity.deadline.toDateString()}`);
      });
    }
  
    // Get activities that fall within the specified month and year
    getActivitiesInMonth(year, month) {
      return this.activities.filter(activity => {
        const startDate = activity.activityStart;
        const endDate = activity.deadline;
  
        return (startDate.getFullYear() === year && startDate.getMonth() + 1 === month) ||
               (endDate.getFullYear() === year && endDate.getMonth() + 1 === month);
      });
    }
  }
  
  class Activity {
    constructor(name, activityStart, deadline) {
      this.name = name;
      this.activityStart = activityStart;
      this.deadline = deadline;
    }
  }
  
  // Example usage:
  const calendar = new Calendar();
  
  // Creating sample activities
  calendar.createActivity("Create Login Page", 5, 10, 2024, 5, 15, 2024);
  calendar.createActivity("Test", 11, 11, 2024, 11, 11, 2024);
  
  // Displaying the calendar for the entire year of 2024
  calendar.displayCalendarYear(2024);
  
  // Displaying October of 2024
  calendar.displayCalendar(2024, 10);
  